### PR TITLE
DMs: make bulletproof

### PR DESCRIPTION
**Pull request by liam-fitzgerald**
First pass at a simple, forwarding only DM networking implementation.

Backend:
- [x] Forward add-nodes
- [x] Make hark not require a group for graph notifications
- [x] Permissioning for inbox (waiting on design spec)

Frontend:
- [x] Render DMs in sidebar
- [x] Render DM route
- [x] Refactor Chat components to not depend on associations everywhere
- [x] Custom DM settings page

### CODE REVIEW

**Comment by tacryt-socryp** [comment]
This definitely needs to assert that the toplevel indices in the validators are valid @p's. The fastest way is likely just (lth (met 3 i.index) 128) or lte, one of those two

File: pkg/arvo/app/dm-hook.hoon [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    def   ~(. (default-agent this %|) bowl)
+    gra   ~(. graphlib bowl)
+    io    ~(. agentio bowl)
+    pass  pass:io
+::
+++  on-init  
+  :_  this
+  :_  ~
+  %+  poke-our:pass  %graph-store
+  %+  update:cg:gra  now.bowl
```
Comment (tacryt-socryp) [inline pkg/arvo/app/dm-hook.hoon:30]: Love the style of this cage helper

File: pkg/arvo/app/dm-hook.hoon [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    |=  =ship
+    ^-  (list card)
+    =/  =index:store
+      [ship ~]
+    ?:  (check-node-existence:gra [our.bowl %inbox] index)  ~
+    =|  =node:store
+    =:  author.post.node     our.bowl
+        index.post.node      index
+        time-sent.post.node  now.bowl
+      ==
```
Comment (tacryt-socryp) [inline pkg/arvo/app/dm-hook.hoon:73]: dedent

File: pkg/arvo/app/dm-hook.hoon [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    =/  =wire  /dm/(scot %p ship)
+    (~(poke-our pass wire) %graph-store (update:cg:gra (add-node index node)))^~
```
Comment (tacryt-socryp) [inline pkg/arvo/app/dm-hook.hoon:75]: better in tall form

File: pkg/arvo/app/dm-hook.hoon [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+      (update:cg:gra (add-node index node))
+    %+  welp  (add-missing-root ship)
+    :~  (~(poke pass wire) dock cage)
+        (~(poke-our pass wire) %graph-store cage)
+    ==
+    ::
+  ++  incoming-add
+    |=  nodes=(map index:store node:store)
+    ^-  (list card)
+    ?>  =(1 ~(wyt by nodes))
```
Comment (tacryt-socryp) [inline pkg/arvo/app/dm-hook.hoon:98]: good

File: pkg/arvo/app/dm-hook.hoon [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+    %+  snoc  (add-missing-root ship)
+    %+  ~(poke-our pass wire)  %graph-store
+    %+  update:cg:gra  now.bowl
+    :+  %add-nodes  [our.bowl %inbox]
+    %-  ~(gas by *(map index:store node:store))
+    %+  turn  ~(tap by nodes)
+    |=  [=index:store =node:store]
+    ?>  ?=([@ @ ~] index)
+    ?>  ?=(%empty -.children.node)
+    =/  new-index=index:store
```
Comment (tacryt-socryp) [inline pkg/arvo/app/dm-hook.hoon:110]: 110-113 should be concatenated into one/two lines

File: pkg/arvo/app/graph-store.hoon [hunk 6]
```diff
@@ -970,6 +970,26 @@
...
+    :-  ~  :-  ~  :-  %graph-update-1
+    !>  ^-  update:store
+    :+  now.bowl  %add-nodes
+    :-  [ship term]
+    %-  ~(gas by *(map index:store node:store))
+    %+  turn  (tap:orm u.children)
+    |=  [=atom =node:store]
+    ^-  [index:store node:store]
+    :-  index.post.node
+    node(children [%empty ~])
```
Comment (tacryt-socryp) [inline pkg/arvo/app/graph-store.hoon:992]: Looks great

File: pkg/arvo/gen/dm-hook/dm.hoon [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+        [[him=ship contents=(list content) ~] ~]
+    ==
+=*  our  p.beak
+=/  =post  *post
+=:  author.post     our
+    index.post      ~[him now]
+    time-sent.post  now
+    contents.post   contents
+==
+::
```
Comment (tacryt-socryp) [inline pkg/arvo/gen/dm-hook/dm.hoon:15]: we've needed this

File: pkg/arvo/mar/graph/validator/dm.hoon [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+      [@ ~]  `[%message [0 1] %count %none]
+    ==
+  ::
+  --
+++  grab
+  |%
+  ++  noun
+    |=  p=*
+    =/  ip  ;;(indexed-post p)
+    ?>  ?=(?([@ ~] [@ @ ~]) index.p.ip)
```
Comment (tacryt-socryp) [inline pkg/arvo/mar/graph/validator/dm.hoon:18]: Should assert that the `[@ ~]` case is a valid patp

**Review by tacryt-socryp — CHANGES_REQUESTED** [review]
Looks good, but handling the node modification in `%dm-hook` will be tricky to support while maintaining valid signatures.

File: pkg/arvo/app/dm-hook.hoon [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+    gra   ~(. graphlib bowl)
+    io    ~(. agentio bowl)
+    pass  pass:io
+::
+++  on-init  
+  :_  this
+  :_  ~
+  %+  poke-our:pass  %graph-store
+  %+  update:cg:gra  now.bowl
+  :+  %add-graph  [our.bowl %inbox]
```
Comment (tacryt-socryp) [inline pkg/arvo/app/dm-hook.hoon:31]: Let's change the name of this from `[our.bowl %inbox]` to `[our.bowl %dm-inbox]`

File: pkg/arvo/app/dm-hook.hoon (hunk 2 above)
Comment (ohAitch) [reply pkg/arvo/app/dm-hook.hoon:73]: ?
this looks like proper kingside what were you thinking of

File: pkg/arvo/app/dm-hook.hoon (hunk 9 above)
Comment (tacryt-socryp) [reply pkg/arvo/app/dm-hook.hoon:42]: Bump

File: pkg/arvo/app/graph-store.hoon [hunk 10]
```diff
@@ -748,6 +748,29 @@
...
+    ?~  children  [~ ~]
+    :-  ~  :-  ~  :-  %graph-update-2
+    !>  ^-  update:store
+    :+  now.bowl  %add-nodes
+    :-  [ship term]
+    %-  ~(gas by *(map index:store node:store))
+    %+  murn  (tap:orm u.children)
+    |=  [=atom =node:store]
+    ^-  (unit [index:store node:store])
+    ?.  ?=(%& -.post.node)  ~
```
Comment (tacryt-socryp) [inline pkg/arvo/app/graph-store.hoon:769]: This is wrong. We should return deleted posts as well.

**Review by tacryt-socryp — COMMENTED** [review]
Looks good other than the comment on the shallow-children scry.

**Comment by liam-fitzgerald** [comment]
addressed all comments @tacryt-socryp

**Review by tacryt-socryp — APPROVED** [review]

**Comment by matildepark** [comment]
Merge conflict, then we can just merge this.

**Pull request merged**